### PR TITLE
Abort requests whose logits row is fully NaN instead of sampling noise

### LINKED ISSUE #33187: [Bug/Design] `SGLANG_SANITIZE_NAN_LOGITS`: a fully-NaN logits row becomes a uniform-random sample — garbage streamed to clients as normal output; proposal: opt-in per-request abort
## Summary

Following up on #32968 (transient NaN-contaminated logits on Kimi-K3): we enabled `SGLANG_SANITIZE_NAN_LOGITS=1` (#28829) in production to stop device-side assert crashes. It does stop the crashes — but it exposes a failure mode that is arguably **worse than the crash** for real users:

When an **entire logits row** is NaN, `nan_to_num_(nan=-1e30)` maps every lane to the same value. Softmax over a constant row is a **uniform distribution over the full vocabulary** (163,840 lanes for K3), so the sampler draws pure random tokens:

- The output is multilingual BPE salad (CJK/Latin/Cyrillic/code fragments mixed), streamed to the client as a **healthy 200 response** — no error, no signal, indistinguishable from real output at the serving layer.
- P(EOS) ≈ 1/163,840 per step, so generation essentially never terminates before `max_tokens`.
- The NaN'd step's KV is written to the request's KV cache, so **every subsequent step of that request is also corrupted** — the remainder of the response is garbage, matching what we observe.
- Reserved/special tokens appear as literal text (we observed `<|reserved_token_163672|>` in a delivered response — a token that is effectively unreachable under any healthy distribution; its presence is a reliable smoking gun for near-uniform sampling).

This hit a real interactive user session in our deployment. Sample of delivered output (thinking channel; surrounding user context omitted):

```
The049带点 accessingGEO-leading皮肤和怀抱萌萌 LarsMMDD Ptraky anthrop:"撕心观感ronesTester())))
getitem信息网初级阶段白开水捕捉到<Element不是要CDATApads_ACCESS ingred na_Act坐上温开水日后的有大
строй白萝卜—拍里边征兆在基层政策支持短裙.Hex弄虚rement fireplaces中央的angent…
```

## Environment

Same deployment as #32968: `kimi-k3` branch @ `f37baeb52f` + cherry-picks of #32477 and #33003, image lineage `lmsysorg/sglang:kimi-k3-c6ad1f26-20260729`, single-node TP=8 on 8× Blackwell-class GPUs (SM10x), four identical nodes. **Speculative decoding disabled** (relevant: these events are on the plain prefill→first-token-sampling path, independent of DSPARK).

## Observed events

Overnight after enabling the flag fleet-wide: **≥3 full-row NaN events across 3 of 4 nodes** (22:26Z, 23:07Z, 23:34Z), each logging

```
NaN detected in sampler: next_token_logits; values were sanitized before sampling. …
```

on **all 8 TP ranks simultaneously**. The events we correlated with decode logs coincide with a small prefill batch (e.g. a 64-token new sequence, `cached=0`) at first-token sampling. Zero asserts/crashes since enabling the flag (that part works as designed).

Note this count is a **lower bound**: `_AsyncNanWarner` is a process-lifetime one-shot latch ("Logged once; further occurrences are silent"), so a second event on the same worker is invisible. Operationally this also means the flag currently provides almost no telemetry.

## Why neither existing mode is adequate in production

| Mode | Blast radius | Client experience |
|---|---|---|
| default (assert) | whole engine: all in-flight requests die, ~20 min reload for a large model | connection dropped, no explanation |
| `SANITIZE_NAN_LOGITS=1` | zero crash | **garbage silently delivered as normal output**; poisoned KV can persist for the request; no signal to retry |

#28829 deliberately framed these as orthogonal crash/heal axes. What's missing is a third option with a *request-sized* blast radius.

Precedent for the phenotype: #30989 (`!!!!` collapse, NVFP4) and #22087 (garbled multilingual mixing, GDN kernel) — same visible symptom family, both ultimately kernel bugs. Root causes should of course be fixed in kernels; but transient NaN sources keep appearing (#32968 remains open), and the serving layer needs a sane containment mode in the meantime.

## Proposal: `SGLANG_ABORT_ON_NAN_LOGITS` (opt-in third mode)

Full-row NaN → **abort only the affected request(s)** with a retriable 5xx; partial-row NaN → sanitize as today (healthy lanes dominate after `-1e30`).

Sketch (validated against current `kimi-k3` branch code):

1. In `Sampler._preprocess_logits`, before `nan_to_num_`: compute `full_nan_rows = torch.isnan(logits).all(-1)` (gated on the env; zero overhead when off). Sanitize afterwards as today so downstream kernels never see NaN.
2. Carry the row mask to batch-result processing on the **existing** token-id device→host sync — no new GPU↔CPU syncs.
3. Scheduler: for flagged rows, `req.to_finish = FINISH_ABORT("Transient numerical error (NaN logits); please retry", 5xx)` — the same per-request machinery the running-timeout abort already uses.
4. **Cache hygiene** (the part that makes this more than a one-liner):
   - skip the radix-cache insert for the aborted request — otherwise the poisoned KV becomes **shareable via prefix matching** and a retry/prefix-sibling can read NaN KV;
   - correctly release hybrid state (Mamba/KDA) for the aborted request;
   - consider hierarchical-cache write-through interactions (host-cache copies of poisoned pages).
5. Per-event (throttled) logging with rid/batch context, replacing the one-shot latch for this path — this doubles as telemetry for tracking transient NaN sources like #32968.

We have an implementation of this in progress and intend to submit a PR. Filing this first for design feedback: flag naming/default, choice of status code (we lean 503 to match the running-timeout abort), interaction with the overlap scheduler and speculative decoding, and whether partial-row events deserve a counter/log line as well.

### PR DESCRIPTION
**Pull request by aabhimittal**
Motivation
Fixes #33187.
SGLANG_SANITIZE_NAN_LOGITS maps NaN logits to -1e30, which stops the sampling kernels from returning out-of-vocab ids. But for a row that is entirely NaN it also produces a constant row — a uniform distribution over the vocabulary. Sampling that returns a uniformly random token, which means:
	•	the request streams vocabulary noise to the client as a healthy 200, with no error signal anywhere;
	•	EOS probability collapses to ~1/vocab_size per step, so the request does not terminate normally;
	•	the corrupted KV is inserted into the radix tree and can reach unrelated requests through prefix matching.
Sanitization remains the right call for a partial NaN row — the surviving logits still carry real evidence about the next token. The full-row case is the one with no safe recovery.
Modifications
New SGLANG_ABORT_ON_NAN_LOGITS (default off), a third mode with a request-scoped blast radius alongside the existing crash and sanitize behaviors:
	•	Detect (utils/async_probe.py, layers/sampler.py) — detect_full_nan_rows returns a [#seq] bool mask, read in Sampler.forward on the raw model output before _preprocess_logits sanitizes it away. Returns None (zero work) when the env is off.
	•	Carry (layers/logits_processor.py, managers/utils.py) — the mask travels on LogitsProcessorOutput, and its D2H is appended to copy_to_cpu so it rides the existing next_token_ids copy and its copy_done event. No new synchronization.
	•	Abort (managers/scheduler_components/batch_result_processor.py) — _abort_full_nan_logits_reqs runs in both the prefill and decode result paths, before update_finish_state consumes req.to_finish. Each flagged request gets FINISH_ABORT(..., HTTPStatus.SERVICE_UNAVAILABLE) — the same mechanism and status as the existing running-timeout abort, which TokenizerManager._handle_abort_finish_reason already turns into a retriable 503 for both streaming and non-streaming clients.
	•	Cache hygiene — flagged requests also set skip_radix_cache_insert, the existing Req flag meaning “this request’s KV must never be shared” (until now set only for the PD-disaggregation fake-bootstrap case). release_kv_cache then frees the KV without publishing it to the radix tree, and maybe_cache_unfinished_req skips it too, so prefix matching cannot spread the corruption.
	•	Partial-row NaN is unchanged — those rows still carry real evidence about the next token, so they fall through to sanitization.
Supporting changes:
	•	Throttled logging — _AsyncNanWarner latched on the first hit and went silent for the process lifetime, hiding whether a NaN burst was ongoing. It now counts events and warns at most once per interval, reporting how many forward passes each line covers. Detection stays sync-free (device counter → pinned host memory).
	•	Two defensive getattrs dropped — mem_cache/common.py read skip_radix_cache_insert through getattr(req, ..., False) in both maybe_cache_unfinished_req and release_kv_cache. The attribute is assigned unconditionally in Req.__init__, so neither guard was ever needed.
Scope note: the mask is produced by Sampler.forward only, which is reached from ModelRunner.sample — one logits row per request, so it indexes batch.reqs directly (asserted). Speculative workers verify through their own sampling path in speculative/eagle_utils.py and leave the mask unset, so spec decoding keeps today’s sanitize-only behavior. The reporter in #33187 ran with speculative decoding disabled. Extending coverage to the verify path needs stride-aware row→request mapping and is left for a follow-up.
Known limitation: on the prefill/decode path the sampled token is appended to output_ids before update_finish_state consumes the abort, so a streaming client may see one final noise token alongside the abort finish reason. Non-streaming clients get the 503 and no text. Suppressing that token would mean branching in the per-request hot loop; happy to add it if reviewers prefer.
Accuracy Tests
Not applicable. Model outputs change only in the failure case this targets — a request that would have emitted uniformly random tokens now returns 503 instead — and only when the new env var is explicitly enabled. The default leaves existing behavior unchanged.
New unit tests (test/registered/unit/utils/test_abort_on_nan_logits.py, CPU, base-a-test-cpu), run locally against the current branch:

Ran 7 tests in 0.004s

OK

They cover full-vs-partial row detection, the env-off no-op, the sanitize→uniform mechanism the fix exists to prevent, the abort’s status code and cache-insert coupling, the finished/retracted skip, and the one-row-per-request assertion. test_module_state_ratchet and test_legacy_global_ratchet were also re-run and pass.
Speed Tests and Profiling
No steady-state cost when SGLANG_ABORT_ON_NAN_LOGITS is off: detect_full_nan_rows returns None before touching the tensor, and the D2H is skipped. When on, it adds one isnan().all(-1) reduction per sampling step and a [#seq] bool D2H that rides the existing copy event.
The throttled-logging change does add work under SGLANG_SANITIZE_NAN_LOGITS: the previous latch stopped the per-step isnan().any() reduction after the first warning, and it now runs every step. That reduction is over the same tensor nan_to_num_ already scans unconditionally on that path, so it is one extra pass on an opt-in debug path.
Checklist
	•	Format your code according to the Format code with pre-commit. — isort, black, and ruff --select=F401,F821,UP037 all clean over the changed files.
	•	Add unit tests according to the Run and add unit tests. — 7 new CPU tests, passing (see Accuracy Tests).
	•	Update documentation according to Write documentations. — added the SGLANG_ABORT_ON_NAN_LOGITS row to environment_variables.mdx, and corrected the adjacent SGLANG_SANITIZE_NAN_LOGITS default, which the table listed as true while environ.py has EnvBool(False).
	•	Provide accuracy and speed benchmark results according to Test the accuracy and Benchmark the speed. — not applicable, see above.
	•	Follow the SGLang code style guidance.
Review and Merge Process
	1.	Ping Merge Oncalls to start the process. See the PR Merge Process.
	2.	Get approvals from CODEOWNERS and other reviewers.
	3.	Trigger CI tests with comments or contact authorized users to do so.
	•	Common commands include /tag-and-rerun-ci, /tag-run-ci-label, /rerun-failed-ci
	4.	After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #30820614815](https://github.com/sgl-project/sglang/actions/runs/30820614815)
Latest PR Test (Extra): :x: [Run #30820613283](https://github.com/sgl-project/sglang/actions/runs/30820613283)

### CODE REVIEW

**Comment by aabhimittal** [comment]
CI is gated on the run-ci label; could a maintainer add it? All 16 red checks are pr-gate/finish from the missing label, no test has run yet.
